From 8bedfcf9e43aa5af19c561168f073ca554fdbb80 Mon Sep 17 00:00:00 2001
From: Koji Matsuoka <koji.matsuoka.xm@renesas.com>
Date: Mon, 17 Jun 2013 10:54:56 +0900
Subject: [PATCH 0525/1083] ARM: shmobile: r8a7790: Add QSPI driver support

Signed-off-by: Koji Matsuoka <koji.matsuoka.xm@renesas.com>
(cherry picked from commit 37824dde5a3effb5bc6695bb5a4ddb6f87095f14)

Conflicts:
	arch/arm/configs/shmobile_defconfig
	arch/arm/mach-shmobile/clock-r8a7790.c
	arch/arm/mach-shmobile/setup-r8a7790.c

Signed-off-by: Ryo Kataoka <ryo.kataoka.wt@renesas.com>
---
 arch/arm/configs/shmobile_defconfig    | 16 +++++++-
 arch/arm/mach-shmobile/clock-r8a7790.c |  9 +++-
 arch/arm/mach-shmobile/setup-r8a7790.c | 75 ++++++++++++++++++++++++++++++++++
 3 files changed, 98 insertions(+), 2 deletions(-)

diff --git a/arch/arm/configs/shmobile_defconfig b/arch/arm/configs/shmobile_defconfig
index 63fe511..3c169e1 100644
--- a/arch/arm/configs/shmobile_defconfig
+++ b/arch/arm/configs/shmobile_defconfig
@@ -53,6 +53,7 @@ CONFIG_UNIX=y
 CONFIG_INET=y
 CONFIG_IP_PNP=y
 CONFIG_IP_PNP_DHCP=y
+CONFIG_EXT2_FS=y
 # CONFIG_INET_XFRM_MODE_TRANSPORT is not set
 # CONFIG_INET_XFRM_MODE_TUNNEL is not set
 # CONFIG_INET_XFRM_MODE_BEET is not set
@@ -63,14 +64,21 @@ CONFIG_IP_PNP_DHCP=y
 CONFIG_UEVENT_HELPER_PATH="/sbin/hotplug"
 CONFIG_DEVTMPFS=y
 CONFIG_DEVTMPFS_MOUNT=y
+CONFIG_REGMAP_SPI=y
 CONFIG_MTD=y
+CONFIG_MTD_OF_PARTS=y
 CONFIG_MTD_CHAR=y
+CONFIG_MTD_BLKDEVS=y
 CONFIG_MTD_BLOCK=y
 CONFIG_MTD_CFI=y
 CONFIG_MTD_CFI_ADV_OPTIONS=y
 CONFIG_MTD_CFI_INTELEXT=y
 CONFIG_MTD_PHYSMAP=y
 CONFIG_MTD_BLOCK2MTD=y
+CONFIG_MTD_M25P80=y
+CONFIG_M25PXX_USE_QUAD_READ=y
+CONFIG_OF_SPI=y
+CONFIG_OF_MTD=y
 CONFIG_SCSI=y
 CONFIG_BLK_DEV_SD=y
 CONFIG_MD=y
@@ -109,6 +117,9 @@ CONFIG_I2C_SH_MOBILE=y
 CONFIG_GPIO_EM=y
 CONFIG_DRM=y
 CONFIG_DRM_RCAR_DU=y
+CONFIG_SPI=y
+CONFIG_SPI_MASTER=y
+CONFIG_SPI_QSPI=y
 # CONFIG_HWMON is not set
 CONFIG_REGULATOR=y
 CONFIG_MEDIA_SUPPORT=y
@@ -157,7 +168,10 @@ CONFIG_UIO_PDRV_GENIRQ=y
 CONFIG_MSDOS_FS=y
 CONFIG_VFAT_FS=y
 CONFIG_TMPFS=y
-# CONFIG_MISC_FILESYSTEMS is not set
+CONFIG_MISC_FILESYSTEMS=y
+CONFIG_JFFS2_FS=y
+CONFIG_JFFS2_FS_DEBUG=0
+CONFIG_JFFS2_FS_WRITEBUFFER=y
 CONFIG_NFS_FS=y
 CONFIG_NFS_V3_ACL=y
 CONFIG_NFS_V4=y
diff --git a/arch/arm/mach-shmobile/clock-r8a7790.c b/arch/arm/mach-shmobile/clock-r8a7790.c
index f70cce6..5e56d1a 100644
--- a/arch/arm/mach-shmobile/clock-r8a7790.c
+++ b/arch/arm/mach-shmobile/clock-r8a7790.c
@@ -110,6 +110,10 @@ static struct clk hp_clk = {
 	.parent		= &pll1_d2_clk,
 };
 
+static struct clk qspi_clk = {
+	.rate		= 97500000,
+};
+
 static struct clk mp_clk = {
 	.rate		= 52000000,
 	.mapping	= &cpg_mapping,
@@ -127,6 +131,7 @@ static struct clk *main_clks[] = {
 	&pll1_d4_clk,
 	&zg_clk,
 	&hp_clk,
+	&qspi_clk,
 	&mp_clk,
 	&cp_clk,
 };
@@ -174,7 +179,7 @@ enum {
 	MSTP216, MSTP207, MSTP206, MSTP204, MSTP203, MSTP202,
 	MSTP726, MSTP725, MSTP724, MSTP723, MSTP721, MSTP720, MSTP704, MSTP703,
 	MSTP314, MSTP313, MSTP312, MSTP311,
-	MSTP929, MSTP922,
+	MSTP929, MSTP922, MSTP917,
 	MSTP1031, MSTP1030, MSTP1019, MSTP1018, MSTP1017, MSTP1015, \
 	MSTP1014, MSTP1005,
 	MSTP_NR };
@@ -199,6 +204,7 @@ static struct clk mstp_clks[MSTP_NR] = {
 	[MSTP720] = SH_CLK_MSTP32(&cp_clk, SMSTPCR7, 20, 0),
 	[MSTP704] = SH_CLK_MSTP32(&mp_clk, SMSTPCR7, 04, 0),
 	[MSTP703] = SH_CLK_MSTP32(&mp_clk, SMSTPCR7, 03, 0),
+	[MSTP917] = SH_CLK_MSTP32(&qspi_clk, SMSTPCR9, 17, 0),
 	[MSTP929] = SH_CLK_MSTP32(&hp_clk, SMSTPCR9, 29, 0),
 	[MSTP922] = SH_CLK_MSTP32(&hp_clk, SMSTPCR9, 22, 0),
 	[MSTP1031] = SH_CLK_MSTP32(&hp_clk, SMSTPCR10, 31, 0),
@@ -243,6 +249,7 @@ static struct clk_lookup lookups[] = {
 	CLKDEV_DEV_ID("sh-sci.7", &mstp_clks[MSTP720]),
 	CLKDEV_CON_ID("hs_usb", &mstp_clks[MSTP704]),
 	CLKDEV_CON_ID("usb_fck", &mstp_clks[MSTP703]),
+	CLKDEV_DEV_ID("qspi.0", &mstp_clks[MSTP917]),
 	CLKDEV_DEV_ID("i2c-rcar.2", &mstp_clks[MSTP929]),
 	CLKDEV_DEV_ID("adg", &mstp_clks[MSTP922]),
 	CLKDEV_DEV_ID("src0", &mstp_clks[MSTP1031]),
diff --git a/arch/arm/mach-shmobile/setup-r8a7790.c b/arch/arm/mach-shmobile/setup-r8a7790.c
index 09bee62..5f30e9c 100644
--- a/arch/arm/mach-shmobile/setup-r8a7790.c
+++ b/arch/arm/mach-shmobile/setup-r8a7790.c
@@ -35,6 +35,10 @@
 #include <linux/mmc/sh_mobile_sdhi.h>
 #include <linux/regulator/fixed.h>
 #include <linux/regulator/machine.h>
+#include <linux/mtd/mtd.h>
+#include <linux/mtd/partitions.h>
+#include <linux/spi/spi.h>
+#include <linux/spi/flash.h>
 #include <mach/common.h>
 #include <mach/irqs.h>
 #include <mach/r8a7790.h>
@@ -763,6 +767,73 @@ static struct platform_device sdhi3_device = {
 	}
 };
 
+/* QSPI resource */
+static struct resource qspi_resources[] = {
+	[0] = {
+		.name = "QSPI",
+		.start = 0xe6b10000,
+		.end = 0xe6b10fff,
+		.flags = IORESOURCE_MEM,
+	},
+	[1] = {
+		.start = gic_spi(184),
+		.flags = IORESOURCE_IRQ,
+	},
+};
+
+/* SPI Flash memory (Spansion S25FL512SAGMFIG11) */
+static struct mtd_partition spiflash_part[] = {
+	/* Reserved for user loader program, read-only */
+	[0] = {
+		.name = "loader_prg",
+		.offset = 0,
+		.size = SZ_256K,
+		.mask_flags = MTD_WRITEABLE,	/* read only */
+	},
+	/* Reserved for user program, read-only */
+	[1] = {
+		.name = "user_prg",
+		.offset = MTDPART_OFS_APPEND,
+		.size = SZ_4M,
+		.mask_flags = MTD_WRITEABLE,	/* read only */
+	},
+	/* All else is writable (e.g. JFFS2) */
+	[2] = {
+		.name = "flash_fs",
+		.offset = MTDPART_OFS_APPEND,
+		.size = MTDPART_SIZ_FULL,
+		.mask_flags = 0,
+	},
+};
+
+static struct flash_platform_data spiflash_data = {
+	.name		= "m25p80",
+	.parts		= spiflash_part,
+	.nr_parts	= ARRAY_SIZE(spiflash_part),
+	.type		= "s25fl512s",
+};
+
+static struct platform_device qspi_device = {
+	.name = "qspi",
+	.id = 0,
+	.num_resources = ARRAY_SIZE(qspi_resources),
+	.resource	= qspi_resources,
+	.dev  = {
+		.platform_data = &spiflash_data,
+	},
+};
+
+static struct spi_board_info spi_info[] __initdata = {
+	{
+		.modalias		= "m25p80",
+		.platform_data		= &spiflash_data,
+		.mode			= SPI_MODE_0,
+		.max_speed_hz		= 30000000,
+		.bus_num		= 0,
+		.chip_select		= 0,
+	},
+};
+
 static struct platform_device *r8a7790_early_devices[] __initdata = {
 	&eth_device,
 	&powervr_device,
@@ -777,6 +848,7 @@ static struct platform_device *r8a7790_early_devices[] __initdata = {
 	&sdhi1_device,
 	&sdhi2_device,
 	&sdhi3_device,
+	&qspi_device,
 };
 
 static struct renesas_irqc_config irqc0_data = {
@@ -831,6 +903,9 @@ void __init r8a7790_add_standard_devices(void)
 			     ARRAY_SIZE(r8a7790_early_devices));
 
 	r8a7790_add_device_to_domain(&r8a7790_rgx, &powervr_device);
+
+	/* QSPI flash memory */
+	spi_register_board_info(spi_info, ARRAY_SIZE(spi_info));
 }
 
 void __init r8a7790_timer_init(void)
-- 
1.8.3.2

